### PR TITLE
Put strict strong name checking behind a configure flag and runtime option

### PR DESCRIPTION
**Pull request by lambdageek**
* `--enable-minimal=desktop_loader` configure flag that replaces `--enable-minimal=assembly_remapping` and sets `DISABLE_DESKTOP_LOADER` define (replacing `DISABLE_ASSEMBLY_REMAPPING`).
* silently we keep accepting `assembly_remapping` as an alias of `desktop_loader`.
* Add `--assembly-loader=[strict,legacy]` command line option.  Default: strict.  Relevant only if `!defined(DISABLE_DESKTOP_LOADER)`
* If strict we will refuse to load any assembly in application base or MONO_PATH that has the right file name but doesn't match the public key token and version number.  If legacy, use old mono behavior (if we find a file with the right name, go with it).

### CODE REVIEW

File: mono/mini/driver.c [hunk 1]
```diff
@@ -1941,6 +1941,14 @@ mono_main (int argc, char* argv[])
 		} else if (strcmp (argv [i], "--nacl-null-checks-off") == 0){
 			nacl_null_checks_off = TRUE;
 #endif
+		} else if (strncmp (argv [i], "--assembly-loader=", strlen("--assembly-loader=")) == 0) {
```
Comment (xmcclure) [inline mono/mini/driver.c]: One nice thing about this approach is we could add a third "warn" option at some point that emits stderr spew every time it detects we're loading an assembly of the wrong strongname.

**Comment by xmcclure** [comment]
I think that the default for DISABLE_DESKTOP_LOADER should be changed to "no" and the default for "assembly-loader" should be changed to "legacy" (in other words, it should be built in by default at configure time, but off by default at run time). Otherwise looks good.

Also, at some point we need tests and an explanation in `mono.1`.

**Comment by xmcclure** [comment]
(Also names are really not important but I would suggest naming the macro "standard loader" or "strict loader" because "desktop" does not communicate much. Among other things, the implication with the current naming seems to be that Cocoa is not "desktop".)

**Comment by lambdageek** [comment]
Per @kumpera's request off-github, DISABLE_DESKTOP_LOADER is coupled with assembly remapping.  It doesn't make sense to disable assembly remapping on desktop.  Think of DISABLE_DESKTOP_LOADER as "I put all my assemblies in one directory, leave me alone".

We can debate `--assembly-loader` default on the mailing list.

Some tests went in on previous commits.

File: mono/metadata/assembly.c [hunk 2]
```diff
@@ -89,7 +89,7 @@ static char **assemblies_path = NULL;
 /* Contains the list of directories that point to auxiliary GACs */
 static char **extra_gac_paths = NULL;
 
-#ifndef DISABLE_ASSEMBLY_REMAPPING
+#ifndef DISABLE_DESKTOP_LOADER
```
Comment (marek-safar) [inline mono/metadata/assembly.c]: Wouldn't be better instead of two negatives to have simple `#if DESKTOP_LOADER_FEATURE` or something similar?

Comment (lambdageek) [inline mono/metadata/assembly.c]: Perhaps.  I was trying to stay consistent with all the other `--enable-minimal` flags which all define `DISABLE_feature` if that particular subsystem is excluded.  I thought that was the idiom in the runtime.

Comment (kumpera) [inline mono/metadata/assembly.c]: I think keeping it consistent with the rest of the enable-minimal is better. There's enough of it around the runtime that we're used to it.

**Comment by kumpera** [comment]
LGTM

**Pull request merged**